File: macos/Classes/SignalR.swift
//
//  SignalR.swift
//
//  Created by Arash on 4/9/22.
//

import Foundation

class SignalR {
    static let instance = SignalR()
    
    var hubProxy: HubProxy?
    var hubConnection: HubConnection?
    
    func connect(url: String, hubName: String, eventName: String, queryString: [String: String], updateConnectionStatus: @escaping (Int) -> Void, newMessage: @escaping (Dictionary<String, Any>) -> Void) {
        hubConnection = HubConnection(withUrl: url)
        hubProxy = hubConnection!.createHubProxy(hubName: hubName)
        
        hubConnection!.queryString = queryString
        
        _ = hubProxy!.on(eventName: eventName) { args in
            if let dictionary = args[0] as? Dictionary<String, Any> {
                newMessage(dictionary)
            }
        }
        
        // MARK: - SignalR events
        hubConnection!.started = {
            updateConnectionStatus(ConnectionStatus.connected.rawValue)
        }

        hubConnection!.reconnecting = {
            updateConnectionStatus(ConnectionStatus.reconnecting.rawValue)
        }

        hubConnection!.reconnected = {
            updateConnectionStatus(ConnectionStatus.reconnected.rawValue)
        }

        hubConnection!.closed = {
            updateConnectionStatus(ConnectionStatus.disconnected.rawValue)
        }

        hubConnection!.connectionSlow = {
            updateConnectionStatus(ConnectionStatus.connectionSlow.rawValue)
        }

        hubConnection!.error = { error in
            print("Connection error", error)
            
            updateConnectionStatus(ConnectionStatus.connectingError.rawValue)
        }
        
        hubConnection!.start()
        
        updateConnectionStatus(ConnectionStatus.connecting.rawValue)
    }
    
    func disconnect() {
        hubConnection?.stop()
    }
    
    func isConnected() -> Bool {
        return hubConnection?.state == SignalRConnectionState.connected
    }
}

enum ConnectionStatus: Int {
    case connected = 0
    case connecting = 1
    case disconnected = 2
    case reconnected = 3
    case reconnecting = 4
    case connectionSlow = 5
    case connectingError = 6
}

File: android/src/main/kotlin/com/arashz4/signalr_socket/SignalR.kt
package com.arashz4.signalr_socket

import android.os.Handler
import android.os.Looper
import com.github.signalr4j.client.ConnectionState
import com.github.signalr4j.client.LogLevel
import com.github.signalr4j.client.hubs.HubConnection

object SignalR {
    private lateinit var connection: HubConnection

    fun connect(url: String, hubName: String, eventName: String, queryStrings: Map<String, String>, updateStatus: (Int) -> Unit, newMessage: (Any) -> Unit, connected: () -> Unit) {
        try {
            var queryString = ""
            for (query in queryStrings) {
                if (queryString.isEmpty()) {
                    queryString = query.key + "=" + query.value
                }
                else {
                    queryString += "?" + query.key + "=" + query.value
                }
            }

            connection = HubConnection(url, queryString, false) { _: String, _: LogLevel -> }
            val hub = connection.createHubProxy(hubName)

            hub.on(eventName, { response ->
                Handler(Looper.getMainLooper()).post {
                    newMessage(response)
                }
            }, Any::class.java)

            var isFirst = true

            connection.connected {
                Handler(Looper.getMainLooper()).post {
                    if (isFirst) {
                        isFirst = false
                        connected()
                    }
                    updateStatus(ConnectionStatus.Connected.value)
                }
            }

            connection.reconnected {
                Handler(Looper.getMainLooper()).post {
                    updateStatus(ConnectionStatus.Reconnected.value)
                }
            }

            connection.reconnecting {
                Handler(Looper.getMainLooper()).post {
                    updateStatus(ConnectionStatus.Reconnecting.value)
                }
            }

            connection.closed {
                Handler(Looper.getMainLooper()).post {
                    updateStatus(ConnectionStatus.Disconnected.value)
                }
            }

            connection.connectionSlow {
                Handler(Looper.getMainLooper()).post {
                    updateStatus(ConnectionStatus.ConnectionSlow.value)
                }
            }

            connection.error {
                Handler(Looper.getMainLooper()).post {
                    updateStatus(ConnectionStatus.Error.value)
                }
            }

            connection.start()
        } catch (ex: Exception) {
            println("Error " + ex.localizedMessage)
        }
    }

    fun disconnect(disconnected: () -> Unit) {
        try {
            if (this::connection.isInitialized) {
                connection.stop()
                disconnected()
            }
        } catch (ex: Exception) {
            println(ex.localizedMessage?.plus(ex.stackTrace.toString()))
        }
    }

    fun isConnected(state: (Boolean) -> Unit) {
        try {
            if (this::connection.isInitialized) {
                state(connection.state == ConnectionState.Connected)
            } else {
                state(false)
            }
        } catch (ex: Exception) {
            println("Error " + ex.localizedMessage)
        }
    }
}

enum class ConnectionStatus(val value: Int) {
    Connected(0),
    Disconnected(1),
    Reconnected(2),
    Reconnecting(3),
    ConnectionSlow(4),
    Error(5)
}